### DISCUSSION TITLE
The peak of memory consumption is too high.

### DISCUSSION THREAD

**Discussion by PeizeLin** [deepmodeling/abacus-develop#2302] [General]
### Describe the bug

Two layers of graphene, 1588 C atoms.
PBE, scf.

Between LCAO_gen_fixedH::b_NL_beta_new and OperatorLCAO::init, the memory consumption raises from 40GB to 218GB then reduces to 43GB.
![image](https://user-images.githubusercontent.com/78645006/221580657-e167a024-aa01-4265-8a92-ec46f600ef52.png)

Between Gint_interface::cal_gint_vlocal and OperatorLCAO::folding_fixed, the memory consumption raises from 43GB to 221GB then reduces to 56GB.
![image](https://user-images.githubusercontent.com/78645006/221580707-2104cfe6-03f7-4187-95f2-e66aa21c27f5.png)

In the other parts of program, the memory consumption keeps low.
So these two peak limits the calculation of big molecules.

[C_11_k1_mpi1_4.zip](https://github.com/deepmodeling/abacus-develop/files/10840440/C_11_k1_mpi1_4.zip)

### Environment

Linux 3.10.0-1160.el7.x86_64, Red Hat 4.8.5-44
icpc 2021.5.0 (gcc 10.2.0)
intelmpi 2021.5 / mpich 4.1
mkl 2021.5
elpa_openmp 2021.11.002
cereal 1.3.2

**Comment by dyzheng** [comment]
build_Nonlocal_beta_new() will cost too much temporary memory for multi threads calculation, maybe parallel by atoms is not a suitable scheme. Can you look at this problem @Alcanderian ?

**Comment by Alcanderian** [comment]
Yes, we can consider another multithreading scheme.
Before that, I need to examine the problem size of the inner loop.
And maybe we should switch scheme among different problem size.
